File: backend/api/ipfs.py
"""
IPFS API endpoints for pinning and retrieving encrypted diary data

All data is encrypted client-side before pinning to IPFS
This ensures privacy even on public IPFS network
"""

from fastapi import APIRouter, HTTPException, status
from pydantic import BaseModel, Field
import base64
import os
import httpx
from typing import Optional, Dict, Any


router = APIRouter(prefix="/api/ipfs", tags=["ipfs"])


# Configuration
IPFS_API_URL = os.getenv("IPFS_API_URL", "http://127.0.0.1:5001")
IPFS_GATEWAY = os.getenv("IPFS_GATEWAY", "https://ipfs.io/ipfs/")
MAX_FILE_SIZE = 10 * 1024 * 1024  # 10MB max


# Request/Response models
class PinRequest(BaseModel):
    data: str = Field(..., description="Base64-encoded encrypted data")
    metadata: Optional[Dict[str, Any]] = Field(None, description="Optional metadata (not encrypted)")


class PinResponse(BaseModel):
    cid: str = Field(..., description="IPFS Content Identifier")
    uri: str = Field(..., description="IPFS URI (ipfs://...)")
    size: int = Field(..., description="Size in bytes")


class RetrieveResponse(BaseModel):
    data: str = Field(..., description="Base64-encoded encrypted data")
    size: int = Field(..., description="Size in bytes")


class IPFSHealthResponse(BaseModel):
    ipfs_available: bool
    api_url: str
    gateway: str


# Helper functions
def is_ipfs_available() -> bool:
    """Check if IPFS node is available"""
    try:
        response = httpx.get(f"{IPFS_API_URL}/api/v0/version", timeout=2.0)
        return response.status_code == 200
    except Exception:
        return False


async def pin_to_ipfs(data: bytes) -> str:
    """
    Pin data to IPFS node

    Args:
        data: Raw bytes to pin

    Returns:
        IPFS CID (Content Identifier)
    """
    try:
        # Use IPFS HTTP API
        files = {"file": data}
        async with httpx.AsyncClient() as client:
            response = await client.post(
                f"{IPFS_API_URL}/api/v0/add",
                files=files,
                params={"pin": "true"},  # Pin by default
                timeout=30.0
            )

        if response.status_code != 200:
            raise HTTPException(
                status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
                detail=f"IPFS add failed: {response.text}"
            )

        result = response.json()
        return result["Hash"]

    except httpx.TimeoutException:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="IPFS node timeout"
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail=f"IPFS error: {str(e)}"
        )


async def retrieve_from_ipfs(cid: str) -> bytes:
    """
    Retrieve data from IPFS

    Args:
        cid: IPFS Content Identifier

    Returns:
        Raw bytes from IPFS
    """
    try:
        async with httpx.AsyncClient() as client:
            response = await client.post(
                f"{IPFS_API_URL}/api/v0/cat",
                params={"arg": cid},
                timeout=30.0
            )

        if response.status_code == 200:
            return response.content
        elif response.status_code == 404:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"CID not found: {cid}"
            )
        else:
            raise HTTPException(
                status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
                detail=f"IPFS cat failed: {response.text}"
            )

    except httpx.TimeoutException:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="IPFS node timeout"
        )
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail=f"IPFS error: {str(e)}"
        )


# Endpoints
@router.post("/pin", response_model=PinResponse)
async def pin_encrypted_data(request: PinRequest):
    """
    Pin encrypted data to IPFS

    **Privacy:** Data must be encrypted client-side before calling this endpoint.
    Only encrypted blobs are pinned to IPFS, never plaintext.

    **Size Limit:** 10MB per file

    Returns IPFS CID and URI for later retrieval.
    """
    # Decode base64 data
    try:
        data_bytes = base64.b64decode(request.data)
    except Exception:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid base64 data"
        )

    # Check size limit
    if len(data_bytes) > MAX_FILE_SIZE:
        raise HTTPException(
            status_code=status.HTTP_413_REQUEST_ENTITY_TOO_LARGE,
            detail=f"Data exceeds max size of {MAX_FILE_SIZE} bytes"
        )

    # Check IPFS availability
    if not is_ipfs_available():
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="IPFS node not available"
        )

    # Pin to IPFS
    cid = await pin_to_ipfs(data_bytes)

    return PinResponse(
        cid=cid,
        uri=f"ipfs://{cid}",
        size=len(data_bytes)
    )


@router.get("/{cid}", response_model=RetrieveResponse)
async def retrieve_encrypted_data(cid: str):
    """
    Retrieve encrypted data from IPFS

    **Privacy:** Returns encrypted data. Client must decrypt with their key.

    Args:
        cid: IPFS Content Identifier

    Returns base64-encoded encrypted data.
    """
    # Check IPFS availability
    if not is_ipfs_available():
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="IPFS node not available"
        )

    # Retrieve from IPFS
    data_bytes = await retrieve_from_ipfs(cid)

    # Encode to base64 for JSON response
    data_base64 = base64.b64encode(data_bytes).decode()

    return RetrieveResponse(
        data=data_base64,
        size=len(data_bytes)
    )


@router.get("/health", response_model=IPFSHealthResponse)
async def ipfs_health():
    """
    Check IPFS node availability

    Returns IPFS status and configuration.
    """
    available = is_ipfs_available()

    return IPFSHealthResponse(
        ipfs_available=available,
        api_url=IPFS_API_URL,
        gateway=IPFS_GATEWAY
    )

File: frontend/lib/ipfs.ts
/**
 * IPFS Integration for Frontend
 *
 * Provides utilities to pin encrypted diary data to IPFS via backend API
 * Retrieves and decrypts data from IPFS
 *
 * Privacy: All data is encrypted client-side before pinning
 */

import { encryptAndSerialize, decryptSerialized } from './encryption';
import { getMasterKey } from './keyManager';

const API_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000';

export interface IPFSPinResult {
  cid: string;
  uri: string;
  size: number;
}

export interface IPFSRetrieveResult {
  data: string; // Encrypted data (base64)
  size: number;
}

/**
 * Check if IPFS is available via backend
 *
 * @returns True if IPFS node is available
 */
export async function isIPFSAvailable(): Promise<boolean> {
  try {
    const response = await fetch(`${API_URL}/api/ipfs/health`);
    if (!response.ok) return false;

    const data = await response.json();
    return data.ipfs_available === true;
  } catch (error) {
    console.error('IPFS health check failed:', error);
    return false;
  }
}

/**
 * Pin encrypted data to IPFS
 *
 * Data is encrypted client-side before sending to backend
 *
 * @param plaintext - Data to encrypt and pin
 * @param metadata - Optional metadata (not encrypted)
 * @returns IPFS CID and URI
 */
export async function pinToIPFS(
  plaintext: string,
  metadata?: Record<string, unknown>
): Promise<IPFSPinResult> {
  // Encrypt data client-side
  const key = await getMasterKey();
  const encrypted = await encryptAndSerialize(plaintext, key);

  // Send encrypted data to backend
  const response = await fetch(`${API_URL}/api/ipfs/pin`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      data: btoa(encrypted), // Base64 encode for transport
      metadata,
    }),
  });

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.detail || 'Failed to pin to IPFS');
  }

  const result: IPFSPinResult = await response.json();
  return result;
}

/**
 * Retrieve and decrypt data from IPFS
 *
 * @param cid - IPFS Content Identifier
 * @returns Decrypted plaintext
 */
export async function retrieveFromIPFS(cid: string): Promise<string> {
  // Retrieve encrypted data from backend
  const response = await fetch(`${API_URL}/api/ipfs/${cid}`);

  if (!response.ok) {
    if (response.status === 404) {
      throw new Error('Data not found on IPFS');
    }
    const error = await response.json();
    throw new Error(error.detail || 'Failed to retrieve from IPFS');
  }

  const result: IPFSRetrieveResult = await response.json();

  // Decode base64
  const encrypted = atob(result.data);

  // Decrypt client-side
  const key = await getMasterKey();
  const plaintext = await decryptSerialized(encrypted, key);

  return plaintext;
}

/**
 * Pin diary entry to IPFS
 *
 * Convenience function that encrypts and pins a complete diary entry
 *
 * @param entry - Diary entry object
 * @returns IPFS URI
 */
export async function pinDiaryEntry(entry: {
  goalId: string;
  reflection: string;
  imageDataUrl?: string;
  timestamp: number;
}): Promise<string> {
  // Serialize entry to JSON
  const entryJSON = JSON.stringify(entry);

  // Pin to IPFS
  const result = await pinToIPFS(entryJSON, {
    goalId: entry.goalId,
    timestamp: entry.timestamp,
  });

  return result.uri;
}

/**
 * Retrieve diary entry from IPFS
 *
 * @param cid - IPFS Content Identifier
 * @returns Decrypted diary entry
 */
export async function retrieveDiaryEntry(cid: string): Promise<{
  goalId: string;
  reflection: string;
  imageDataUrl?: string;
  timestamp: number;
}> {
  // Retrieve and decrypt
  const plaintext = await retrieveFromIPFS(cid);

  // Parse JSON
  const entry = JSON.parse(plaintext);

  return entry;
}

/**
 * Convert IPFS URI to CID
 *
 * @param uri - IPFS URI (ipfs://...)
 * @returns CID
 */
export function uriToCID(uri: string): string {
  if (uri.startsWith('ipfs://')) {
    return uri.substring(7); // Remove 'ipfs://' prefix
  }
  return uri;
}

/**
 * Convert CID to IPFS URI
 *
 * @param cid - IPFS Content Identifier
 * @returns IPFS URI
 */
export function cidToURI(cid: string): string {
  if (cid.startsWith('ipfs://')) {
    return cid;
  }
  return `ipfs://${cid}`;
}

/**
 * Get IPFS gateway URL for CID
 *
 * @param cid - IPFS Content Identifier
 * @param gateway - Gateway URL (default: ipfs.io)
 * @returns Full gateway URL
 */
export function getGatewayURL(
  cid: string,
  gateway: string = 'https://ipfs.io/ipfs/'
): string {
  const actualCID = uriToCID(cid);
  return `${gateway}${actualCID}`;
}
